### PR TITLE
Use Django's Javascript catalog feature to manage translatable strings in Javascript

### PR DESCRIPTION
**Pull request by kaedroho**
This PR enables Django's JavaScript catalog for wagtail admin, which provides a global ``gettext()`` function for JavaScript and updates all code in ``client`` to use that instead of manually passing translated strings to JavaScript with a dictionary.

This method is being used on [wagtail-localize](https://github.com/wagtail/wagtail-localize)

String extraction from javascript is done using a ``gettext-extractor`` npm package and this writes the source strings to a separate PO file in the wagtailadmin app called ``djangojs.po``.

See https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#internationalization-in-javascript-code

To test, I'd recommend copying an existing translated wagtailadmin ``django.po`` file to ``djangojs.po`` then run ``compilemessages``. For example, for French

 - ``cp wagtail/admin/locale/fr/LC_MESSAGES/{django,djangojs}.po``
 - ``django-admin compilemessages``

### CODE REVIEW

**Comment by lb-** [comment]
Hey @kaedroho this looks epic, I have a few questions though. 

- Is there any way to use this without adding more functions to the global `window` ? Adding another five global's makes it harder to write parts of code in isolation. 
- Similar to above, but can this be somehow imported into most code? Even if that import is just a simple file that exports references to the global?
- For JS unit tests, is there going to be a global default mock that gets added so that unit tests don't need to set this up each time used?
- For React code, maybe the translation utils can be provided as a component, this avoids pre-resoving the translations until the components actually render (maybe a slight performance improvement) and again helps for many shallow rendering test cases where we don't really care to mock the result of a translation, just that the right bits were passed to our translation component. 

My main concern with adding more globals is that these are hard to discover, if they were all in a namespace like window.wagtail or similar it may be easito work with when we need to.

**Comment by kaedroho** [comment]
Thanks for the quick response @lb-!

> Is there any way to use this without adding more functions to the global window ? Adding another five global's makes it harder to write parts of code in isolation.

There is a [``JSONCatalog`` view](https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#the-jsoncatalog-view) that we could use instead, this gives us access to the strings but we would need to implement our own ``gettext`` function (which we could probably copy from Django's ``JavaScriptCatalog`` implementation).

> For JS unit tests, is there going to be a global default mock that gets added so that unit tests don't need to set this up each time used?

I've added these functions to our stubs file.

> For React code, maybe the translation utils can be provided as a component, this avoids pre-resoving the translations until the components actually render (maybe a slight performance improvement) and again helps for many shallow rendering test cases where we don't really care to mock the result of a translation, just that the right bits were passed to our translation component.

I'm not sure if there would be any noticeable performance improvement from this as ``gettext()`` does a dictionary lookup. The performance doesn't feel any different from testing.

**Comment by lb-** [comment]
@kaedroho thanks for the responses, happy to do a full review if you want (just request a review once those items are updated).

I agree about the performance thing, probably does not make much difference and could risk the catalog system not accessing some translation usages, plus using the function directly does make it easier to understand.

**Comment by lb-** [comment]
@kaedroho just took a look at this and it appears that it adds the functions to a global `django` this may be a good option to access these utils. 

I still think code in client should abstract this to a util that is imported (easier to unit test) 

https://github.com/django/django/blob/7119f40c9881666b6f9b5cf7df09ee1d21cc8344/django/views/i18n.py#L89

**Comment by kaedroho** [comment]
Thanks @lb- I've created a module under `utils` that re-exports these functions from Django's global variable

**Comment by lb-** [comment]
@kaedroho  epic - I will take a look at this.

**Comment by lb-** [comment]
We will need to consider and maybe document how we will support other Django packages also using the same feature. 

For example these docs (although support is for Django 2.2 only) reference a way to extend existing JS translations. 

https://django-recurrence.readthedocs.io/en/latest/installation.html#set-up-internationalization

**Comment by kaedroho** [comment]
> We will need to consider and maybe document how we will support other Django packages also using the same feature.

I think this should be fine as they can also register ``jsi18n`` views and the javascript that Django provides will combine the two: https://github.com/django/django/blob/7119f40c9881666b6f9b5cf7df09ee1d21cc8344/django/views/i18n.py#L110-L116

**Review by lb- — APPROVED** [review]
I think this looks good but as this is a big change to how we do translations, would be good to get other's thoughts.

There is a small size increase to the built JS assets, 0.07MB (unzipped) + 3.67 KB (3.26 KB size) for the `admin/jsi18n/` view.

It's possible we can optimise this slightly with some webpack `dependOn` or chunk management but not sure how to do this myself.

**Comment by lb-** [comment]
@kaedroho just a quick question, when these translation strings show up in Transifex - will the filenames used be provided to translators?

Context is - we want to help ensure the translators have the most helpful info possible when doing this, ensuring we do not 'lose' info from the current approach.

Also, do you know if there is any way to support 'translator comments' in the JavaScript translation system - similar to https://docs.djangoproject.com/en/4.0/topics/i18n/translation/#translator-comments-in-templates

**Comment by lb-** [comment]
just adding design decision needed - I think this looks good but probably need some other input from core team as this is a larger change.

File: client/src/components/Sidebar/modules/MainMenu.tsx [hunk 1]
```diff
@@ -65,7 +67,6 @@ interface MenuProps {
   slim: boolean;
   expandingOrCollapsing: boolean;
   currentPath: string;
-  strings: Strings;
```
Comment (thibaudcolas) [inline client/src/components/Sidebar/modules/MainMenu.tsx:68]: Looks like we’ll also need to remove that prop from the corresponding test cases, and Storybook.

File: package.json [hunk 2]
```diff
@@ -73,6 +73,7 @@
     "enzyme-to-json": "^3.6.2",
     "eslint": "^8.5.0",
     "expose-loader": "^3.1.0",
+    "gettext-extractor": "^3.5.3",
```
Comment (thibaudcolas) [inline package.json:76]: We’ll also need the corresponding `package-lock.json` update.

**Review by thibaudcolas — APPROVED** [review]
Gamechanger :) There are a few things I think still need tweaking, but this PR is prone to conflicts so I’ll merge it with only the most important changes and we can pick up the rest later.

**Pull request merged**